tests/integration/helpers/express-auth.ts

```diff
@@ -0,0 +1,33 @@
+import { okAsync } from 'neverthrow'
```

Comment:
I'm not sure if it's a good practice for helper files to have their own mocks and `expect`s. I can definitely see our tests getting very very messy if we repeat this pattern.

Reply:
Not sure how else to generate an authenticated session besides pinging our own endpoints to authenticate -- the `expects` are to guarantee that we have indeed logged in. Any ideas welcomed!

Reply:
hmmmmm should it just be a helper function in `user.routes.spec.ts`? open to discussion

Reply:
Will have lots of duplication leh, since many other routes are also restricted and require authentication. Can imagine for those routes just need to call this one function to receive a "logged-in" session, that's what this `express-auth` file is for

Reply:
what's the behaviour of the mocks if this file is imported? will the functions remain mocked in the module which imports this file?

Reply:
I don't know hmm; can we cross that bridge when we get to it? Or is that the beginnings of a tech debt :P

Reply:
possible to manually add the cookie and the Session document instead?

Reply:
idk if I'm making too big a deal out of this haha but yes it seems like the beginning of tech debt

Reply:
Not sure if it's possible, but i don't think it's tech debt leh (if the mock doesn't pollute)

Reply:
hm, but if it does pollute then do you think that it's bad practice? I can't find a clear answer online, but I have a hunch that it does pollute, given how I've seen mocks work in other tests.

anyway I don't think it's that big of a deal, and I'm willing to let it go, but with the knowledge that some poor developer will probably have to clean this up in future :P

Reply:
Also not really la, since jest runs each test module in their own sandbox; is just have chance to pollute tests in other tests in the own block. I can mitigate this by unmocking in the helper file directly after the needed mocks are performed!!! (Just thought of this now lol)

Reply:
I still don't like the coupling of all integration tests to `/auth/sendotp` but okay haha will defer to you

Reply:
I've looked around at all integration tests and how they log in; all uses their login route so that the correct cookies are set by `express-session` in the testing environment

Reply:
sure, I agree that if everyone else does it a certain way, then that's a compelling reason for us to do it the same way. but I'm trying to understand why this is the right approach. the first principle of good testing is isolation, right? and integration tests should be isolated in their own modules, just as unit tests are isolated for individual functions? so why is it correct to couple multiple modules to /auth/signout?

also, is it not as easy as using the express-session middleware in tests, and adding a helper middleware which sets req.session.user? why doesn't that work?

Reply:
pinging @arshadali172 and @pallani for thoughts/guidance since yr is on leave! for context, the question is "is it okay for us to use our login endpoint to set the login cookie in integration tests, even if the module being tested is unrelated"

Reply:
also for context this is a big discussion about a small nitpick lol I'm just trying to understand

Reply:
@LoneRifle you might be the best person to answer this.

Reply:
In the context of integration tests, the guiding principle that should prevail is "keep the tests as realistic as possible". It's inevitable that you would define a macro that exercises the login endpoint in this manner before proceeding with the test proper. To that end I would argue that @karrui could have perhaps gone slightly further and mocked the _dependencies_ of MailService and SendOtp, rather than the modules themselves.

The one disadvantage of having a test-wide macro is potential pollution as @mantariksh points out, so you will have to make a habit of resetting that for each suite the macro finds itself in, if indeed the state of the macro is resetting. While not ideal, I think the readability you get from having an instantly-identifiable macro within the tests is worth it.

Note that this is different from unit tests, where you want to keep them as _focused_ as possible, and so, you would keep your helper macros as simple as they can be so ie, if this were a helper for a unit test I would have gone with a smaller macro that just injects state into the user session.

Reply:
thank you for explaining @LoneRifle! if the guiding principle is being realistic then yup, I agree that this is the way to go